### ISSUE TITLE
Autonomous systems page performance

### ISSUE THREAD

**Issue by Queeq** [ntop/ntopng#486]
Hi,
I'm running ntopng (Community v.2.2.160411) on a VMWare virtual host (8 CPUs, 8G RAM), Ubuntu 14.04, custom 4.4.6 kernel. Our external traffic during the day is around 25 Gbit/s. Flows exported from border routers are sampled 4096:1.
Hosts, Autonomous Systems and Country pages from "Hosts" menu take a lot of time to load. I understand that they may need to aggregate a lot of data, but is it okay for it to take tens of seconds to load? If more than one user accesses the system simultaneously it becomes totally unusable for all of them.

The following are scenarios I've tried so far to try to resolve the issue (all within the same virtual machine, local connections over TCP):
1. One nprobe receiving all flows, one ntopng connecting over ZMQ to the nprobe. Observed 100% load on one CPU and <10% load on others. The overloaded CPU was loaded by around 40-60% by Redis server and the rest was system. ntopng was consuming around 40-80% CPU distributed among other CPUs.
2. Six nprobes (one for each FPC carrying uplink traffic on the border router), one ntopng with ZMQ. Pretty much similar picture.
3. Six nprobes + 3 intermediate ntopngs (ZMQ to 2 of the nprobes from each, exporting flows to 1 main ntopng. Separate local Redis server for each of the ntopngs (made this way due to Redis being single-threaded). This way I saw some load distribution among different Redis servers and ntopngs. However main ntopng and the corresponding Redis server kept on eating the majority of CPU (not all though). Subjectively, this lead to some improvement in load time, but it's still big.
4. Same as nr. 3, but six intermediate ntopngs with six Redis servers. The load distributed a bit more. Main ntopng keeps eating 100-150% CPU, main Redis server eats around 30-35% + 30-35% of system load on the same CPU when the page is opened.
5. Setting --dns-mode 3 (i.e. turning it off) makes things better, but it's still unusable by multiple users. Switching between pages on AS page takes more than a minute to complete too.

I/O is not an issue here, disks are not loaded at all.

What bothers me is that Redis uses too small amount of memory while I'd expect it to cache everything. In scenario nr. 5 main instance consumes 6MB (6 megabytes!) and other instances - 12-25 MB. Main ntopng process, however, eats around 900 MB (others - 100-200MB).
In scenario nr. 1 ntopng process consumed small amount of memory, but Redis-server was still keeping small (less than 100 MB).

Something is wrong here.

Now scenario nr. 5 is active. Here's `redis-cli` info for the main process:

```
# Server
redis_version:2.8.4
redis_git_sha1:00000000
redis_git_dirty:0
redis_build_id:a44a05d76f06a5d9
redis_mode:standalone
os:Linux 4.4.6 x86_64
arch_bits:64
multiplexing_api:epoll
gcc_version:4.8.2
process_id:17544
run_id:dd4de3e59ff65e0d4274d04245d857cb88d9bd70
tcp_port:6379
uptime_in_seconds:8196
uptime_in_days:0
hz:10
lru_clock:1342164
config_file:/etc/redis/redis.conf

# Clients
connected_clients:2
client_longest_output_list:0
client_biggest_input_buf:0
blocked_clients:0

# Memory
used_memory:599728
used_memory_human:585.67K
used_memory_rss:6590464
used_memory_peak:1205296
used_memory_peak_human:1.15M
used_memory_lua:33792
mem_fragmentation_ratio:10.99
mem_allocator:jemalloc-3.4.1

# Persistence
loading:0
rdb_changes_since_last_save:0
rdb_bgsave_in_progress:0
rdb_last_save_time:1460455907
rdb_last_bgsave_status:ok
rdb_last_bgsave_time_sec:0
rdb_current_bgsave_time_sec:-1
aof_enabled:0
aof_rewrite_in_progress:0
aof_rewrite_scheduled:0
aof_last_rewrite_time_sec:-1
aof_current_rewrite_time_sec:-1
aof_last_bgrewrite_status:ok

# Stats
total_connections_received:5
total_commands_processed:68594238
instantaneous_ops_per_sec:9501
rejected_connections:0
sync_full:0
sync_partial_ok:0
sync_partial_err:0
expired_keys:3542
evicted_keys:0
keyspace_hits:234749
keyspace_misses:68349836
pubsub_channels:0
pubsub_patterns:0
latest_fork_usec:928

# Replication
role:master
connected_slaves:0
master_repl_offset:0
repl_backlog_active:0
repl_backlog_size:1048576
repl_backlog_first_byte_offset:0
repl_backlog_histlen:0

# CPU
used_cpu_sys:1955.39
used_cpu_user:441.03
used_cpu_sys_children:0.08
used_cpu_user_children:0.02

# Keyspacehad appeared long before that.
db0:keys=86,expires=4,avg_ttl=33774737
```

**Edit 1**
Additionally, in scenario nr. 5 some of the intermediate ntopng processes periodically crash without any messages in their log files. There is stderr message (even though they're run in daemon mode):

```
*** Error in `ntopng': double free or corruption (fasttop): 0x0000000002a03b10 ***
Bad address (stream_engine.cpp:785)
```

**Edit 2**
I figured that the mentioned message about "double free or corruption" has nothing to do with crash. After I've run ntopng instance in foreground, it crashed with the following message:

```
Assertion failed: check () (msg.cpp:220)
Aborted
```

The "double free or corruption" message after it had been running for several minutes, but it had been pretty long time before it crashed (an hour or so).

**Edit 3**
5 out of 6 intermediate ntopng processes crashed over the night. I have noticed two more messages in stderr:

```
Segmentation fault
```

```
Assertion failed: refs_ >= 0 (msg.cpp:272)
```

**Comment by simonemainardi** [comment]
I noticed that you are using version 2.2. There have been huge improvements in sorting and memory management after its release so I encourage you to upgrade to the latest nightly build (version 2.3.x). Please, let us know if the upgrade solved the issue.

BTW, I also think you have to tune ntopng configuration and increase with -X and -x the maximum number of flows and hosts. Make sure actual values are not even close to the maximum ones.

You may also consider using PF_RING zero copy and use RSS to balance either on the nprobe or ntopng side. In this way you can set core affinity and bind each core to a single queue.

**Comment by Queeq** [comment]
I tried installing nightly builds but some of the nprobes were not sending any traffic to ntopng (I used 6 nprobes + 1 ntopng). Performance of the mentioned pages were also quite poor if I remember correctly.

I don't quite understand how `-X` and `-x` options could help. Are there any limits by default that should be risen? I experienced severe problems due to number of flows when I had lower sampling value when the total number of flows rose above 100k. With the current sampling value it is kept relatively low (20-30k).

Regarding PF_RING - as far as I understood it is for local capture and would require traffic mirroring. Not sure if it would be possible as border routers are located in different parts of a country.

 I will test all the things with the latest nightlies once again and report back.

**Comment by simonemainardi** [comment]
Yes, it is worth increasing -X and -x. Of course default settings work quite well in the most part of the cases but very likely you have a huge number of flows/hosts. Increasing X and x will increase internal hash tables and yield less collisions (== faster searches == lower cpu load).

Right PF_RING is only good if local capture is performed. Sorry I missed the part where you say that flows are exported from border routers.

Please, report experience with 2.3.x

**Comment by Queeq** [comment]
I have set both `-x` and `-X` to `1000000`. Web UI performance is indeed much better now. However when I use single ntopng process which receives flows from multiple nprobes it slowly dies after some time: at first it stops showing traffic on separate interfaces, then on all and finally it stops responding completely (it accepts TCP session but never answers for HTTP queries). Should I open separate issue for that?

**Edit 1**
This unresponsiveness happens after around 1 minute 40 seconds uptime, 57k hosts, 47k flows

**Edit 2**
Running in foreground it gives this:

```
15/Apr/2016 15:56:13 [Lua.cpp:4857] WARNING: Script failure [/usr/share/ntopng/scripts/lua/get_grouped_hosts_data.lua][not enough memory]
```

ntopng process consumes around 1.9GB memory at this point. Redis server consumes nothing. There's still around 3GB of free RAM.

**Edit 3**
Sorry, forgot to mention that I'm currently running latest nightly.

**Edit 4**
I left it running and it kept on consuming more and more memory. Killed it after it ate everything.

**Comment by Queeq** [comment]
The AS page performance became much better with v.2.3.160419. Additionally, even though ntopng still freezes after some time, it works longer now: >5 minutes, ~120k hosts and 85k flows. It throws out the same error when running in foreground:

```
19/Apr/2016 11:26:55 [Lua.cpp:4899] WARNING: Script failure [/usr/share/ntopng/scripts/lua/get_grouped_hosts_data.lua][not enough memory]
19/Apr/2016 11:26:55 [Lua.cpp:4899] WARNING: Script failure [/usr/share/ntopng/scripts/lua/get_grouped_hosts_data.lua][not enough memory]
```

I have also noticed that if ntopng is restarted without nprobes being restarted, this new instance doesn't receive traffic from nprobes anymore. Let me know if it is reasonable to open other issue about this.

**Edit**
Inability to receive flows from nprobes is now persistent across ntopng and nprobes restarts.

```
<cut>
19/Apr/2016 12:13:46 [NetworkInterface.cpp:1493] Started packet polling on interface tcp://127.0.0.1:5556 [id: 8]...
19/Apr/2016 12:13:46 [NetworkInterface.cpp:1493] Started packet polling on interface tcp://127.0.0.1:5557 [id: 9]...
19/Apr/2016 12:13:46 [NetworkInterface.cpp:1493] Started packet polling on interface tcp://127.0.0.1:5558 [id: 10]...
19/Apr/2016 12:13:46 [NetworkInterface.cpp:1493] Started packet polling on interface tcp://127.0.0.1:5559 [id: 11]...
19/Apr/2016 12:13:46 [NetworkInterface.cpp:1493] Started packet polling on interface tcp://127.0.0.1:5562 [id: 12]...
19/Apr/2016 12:13:46 [NetworkInterface.cpp:1493] Started packet polling on interface tcp://127.0.0.1:5564 [id: 13]...
19/Apr/2016 12:13:47 [CollectorInterface.cpp:104] Collecting flows on tcp://127.0.0.1:5557 [ntopng->nprobe]
19/Apr/2016 12:13:47 [CollectorInterface.cpp:104] Collecting flows on tcp://127.0.0.1:5556 [ntopng->nprobe]
19/Apr/2016 12:13:47 [CollectorInterface.cpp:104] Collecting flows on tcp://127.0.0.1:5558 [ntopng->nprobe]
19/Apr/2016 12:13:47 [CollectorInterface.cpp:104] Collecting flows on tcp://127.0.0.1:5559 [ntopng->nprobe]
19/Apr/2016 12:13:47 [CollectorInterface.cpp:104] Collecting flows on tcp://127.0.0.1:5562 [ntopng->nprobe]
19/Apr/2016 12:13:47 [CollectorInterface.cpp:104] Collecting flows on tcp://127.0.0.1:5564 [ntopng->nprobe]
19/Apr/2016 12:18:47 [Lua.cpp:4554] WARNING: Script failure [/usr/share/ntopng/scripts/callbacks/minute.lua][not enough memory]
^C19/Apr/2016 12:25:18 [main.cpp:37] Shutting down...
19/Apr/2016 12:25:20 [ProtoStats.cpp:35] [IPv4]  0 B/0.00 Packets
19/Apr/2016 12:25:20 [ProtoStats.cpp:35] [IPv6]  0 B/0.00 Packets
19/Apr/2016 12:25:20 [ProtoStats.cpp:35] [ARP]   0 B/0.00 Packets
19/Apr/2016 12:25:20 [ProtoStats.cpp:35] [MPLS]  0 B/0.00 Packets
19/Apr/2016 12:25:20 [ProtoStats.cpp:35] [Other] 0 B/0.00 Packets
19/Apr/2016 12:25:20 [Ntop.cpp:1252] Interface tcp://127.0.0.1:5556 [running: 0]
19/Apr/2016 12:25:20 [ProtoStats.cpp:35] [IPv4]  0 B/0.00 Packets
19/Apr/2016 12:25:20 [ProtoStats.cpp:35] [IPv6]  0 B/0.00 Packets
19/Apr/2016 12:25:20 [ProtoStats.cpp:35] [ARP]   0 B/0.00 Packets
19/Apr/2016 12:25:20 [ProtoStats.cpp:35] [MPLS]  0 B/0.00 Packets
19/Apr/2016 12:25:20 [ProtoStats.cpp:35] [Other] 0 B/0.00 Packets
19/Apr/2016 12:25:20 [Ntop.cpp:1252] Interface tcp://127.0.0.1:5557 [running: 0]
19/Apr/2016 12:25:20 [ProtoStats.cpp:35] [IPv4]  0 B/0.00 Packets
19/Apr/2016 12:25:20 [ProtoStats.cpp:35] [IPv6]  0 B/0.00 Packets
19/Apr/2016 12:25:20 [ProtoStats.cpp:35] [ARP]   0 B/0.00 Packets
19/Apr/2016 12:25:20 [ProtoStats.cpp:35] [MPLS]  0 B/0.00 Packets
19/Apr/2016 12:25:20 [ProtoStats.cpp:35] [Other] 0 B/0.00 Packets
19/Apr/2016 12:25:20 [Ntop.cpp:1252] Interface tcp://127.0.0.1:5558 [running: 0]
19/Apr/2016 12:25:20 [ProtoStats.cpp:35] [IPv4]  0 B/0.00 Packets
19/Apr/2016 12:25:20 [ProtoStats.cpp:35] [IPv6]  0 B/0.00 Packets
19/Apr/2016 12:25:20 [ProtoStats.cpp:35] [ARP]   0 B/0.00 Packets
19/Apr/2016 12:25:20 [ProtoStats.cpp:35] [MPLS]  0 B/0.00 Packets
19/Apr/2016 12:25:20 [ProtoStats.cpp:35] [Other] 0 B/0.00 Packets
19/Apr/2016 12:25:20 [Ntop.cpp:1252] Interface tcp://127.0.0.1:5559 [running: 0]
19/Apr/2016 12:25:20 [ProtoStats.cpp:35] [IPv4]  0 B/0.00 Packets
19/Apr/2016 12:25:20 [ProtoStats.cpp:35] [IPv6]  0 B/0.00 Packets
19/Apr/2016 12:25:20 [ProtoStats.cpp:35] [ARP]   0 B/0.00 Packets
19/Apr/2016 12:25:20 [ProtoStats.cpp:35] [MPLS]  0 B/0.00 Packets
19/Apr/2016 12:25:20 [ProtoStats.cpp:35] [Other] 0 B/0.00 Packets
19/Apr/2016 12:25:21 [Ntop.cpp:1252] Interface tcp://127.0.0.1:5562 [running: 0]
19/Apr/2016 12:25:21 [ProtoStats.cpp:35] [IPv4]  0 B/0.00 Packets
19/Apr/2016 12:25:21 [ProtoStats.cpp:35] [IPv6]  0 B/0.00 Packets
19/Apr/2016 12:25:21 [ProtoStats.cpp:35] [ARP]   0 B/0.00 Packets
19/Apr/2016 12:25:21 [ProtoStats.cpp:35] [MPLS]  0 B/0.00 Packets
19/Apr/2016 12:25:21 [ProtoStats.cpp:35] [Other] 0 B/0.00 Packets
19/Apr/2016 12:25:21 [Ntop.cpp:1252] Interface tcp://127.0.0.1:5564 [running: 0]
19/Apr/2016 12:25:21 [main.cpp:48] Deleted PID /var/tmp/ntopng.pid [rc: 0]
^C19/Apr/2016 12:29:20 [main.cpp:34] Ok I am leaving now
```

I see normal TCP exchange on the given ports over the loopback interface however.

**Comment by simonemainardi** [comment]
we are working on the improvement of pages performance by performing sorting and grouping operations directly in C. Enhancements will be ready soon. Feel free to open a new issue regarding the nprobes.

**Comment by seismiccollision** [comment]
@Queeq I don't know if this might be any help but at a similar tracked scale using sFlows in ntopng I found that running ntopng with "-S=local" in ntopng.conf helped flush remote hosts (which i didn't care as much about tracking)

here's a snippit from the .conf that's been working for me:
-A=2
-D=all 
-E=all 
-S=all 
-W=3001
-x=3000000
-X=131072
-S=local

**Comment by Queeq** [comment]
Thank you for the suggestion @seismiccollision, I'll try this today.
Just for the sake of statistics, I've left it working overnight and it now shows as consuming 10.4GB of memory (Virtual), 7640 MB residential, so my virtual machine with 8GB RAM has no free memory left and 4GB more was swapped (normally it consumes around 200 MB). There are around 1 090 000 hosts active in ntopng and around 55k flows.

**Comment by seismiccollision** [comment]
@Queeq Did those settings help at all?

**Comment by Queeq** [comment]
@seismiccollision It's difficult to say because my tests failed due to #505. I'll update this thread once it's resolved.

**Comment by Queeq** [comment]
Currently I'm running v.2.3.160504, UI performance is great. Here's the relevant config:

```
# Do not resolve any names
--dns-mode=3

# Limit memory usage
--max-num-flows=200000
--max-num-hosts=250000
--sticky-hosts=none
```

It has been running for 1h 45m now and shows 310k hosts (and rising), 96k flows (and slowly rising), consumes around 3.2-3.6 GB of memory (and seems like rising on average). Sometimes these numbers are getting lower but then rise again.

I've also seen the following message after 1h 35m of uptime, but it looks like it doesn't affect anything:

```
04/May/2016 14:27:09 [Lua.cpp:4958] WARNING: Script failure [/usr/share/ntopng/scripts/lua/get_grouped_hosts_data.lua][not enough memory]
```

I was going to attach it to MySQL so will try `--dump-flows` and `--dump-hosts` after that. I'll post the results here.

**Update**
2h 30m uptime, it shows 150k hosts, 83k flows, memory consumption fluctuates within 3-3.8 GB range.

**Update 2**
2h 33m uptime, several more messages like the following and ntopng stopped responding:

```
04/May/2016 15:26:59 [Lua.cpp:4958] WARNING: Script failure [/usr/share/ntopng/scripts/lua/get_grouped_hosts_data.lua][not enough memory]
04/May/2016 15:27:00 [Lua.cpp:4958] WARNING: Script failure [/usr/share/ntopng/scripts/lua/get_grouped_hosts_data.lua][not enough memory]
04/May/2016 15:27:01 [Lua.cpp:4613] WARNING: Script failure [/usr/share/ntopng/scripts/callbacks/minute.lua][not enough memory]
```

Memory usage - 3.3 GB.

**Comment by Queeq** [comment]
I have configured `--dump-flows` to MySQL database. AS page became a bit slower, but is usable. ntopng memory usage fluctuates between 1.4 and 2.1 GB, flows table in MySQL has around 1.2M rows and using roughly 700 MB of space after ~30 minutes uptime. I've seen no memory warning in ntopng output so far. I will report back after it's run for some more time.

**Comment by Queeq** [comment]
So ntopng had become frozen after 1h 25m of uptime. I haven't been looking at it for another 1.5 hours and now I see it has eaten all RAM and some swap. And I see these messages again:

```
12/May/2016 12:55:07 [Lua.cpp:4962] WARNING: Script failure [/usr/share/ntopng/scripts/lua/get_grouped_hosts_data.lua][not enough memory]
12/May/2016 12:55:07 [Lua.cpp:4962] WARNING: Script failure [/usr/share/ntopng/scripts/lua/get_grouped_hosts_data.lua][not enough memory]
12/May/2016 12:55:07 [Lua.cpp:4962] WARNING: Script failure [/usr/share/ntopng/scripts/lua/get_grouped_hosts_data.lua][not enough memory]
12/May/2016 12:55:07 [Lua.cpp:4962] WARNING: Script failure [/usr/share/ntopng/scripts/lua/get_grouped_hosts_data.lua][not enough memory]
12/May/2016 12:55:07 [Lua.cpp:4962] WARNING: Script failure [/usr/share/ntopng/scripts/lua/get_grouped_hosts_data.lua][not enough memory]
```

I think that an important detail here is that through all this time AS page has been opened in browser.

**Comment by simonemainardi** [comment]
thanks for reporting again. Enhancement has not yet been completed. Please hold on.

**Comment by simonemainardi** [comment]
@Queeq we are done with the improvements. Your issue should now be solved. Please try and update to the latest dev build and give a feedback. Thank you.

**Comment by Queeq** [comment]
Using today's nightly it became frozen after 14 minutes, consuming all available memory.

**Comment by lucaderi** [comment]
Are yiu using the latest 2.3.160601 version? If not please update and report of your setup

**Comment by Queeq** [comment]
@lucaderi Yes, v.2.3.160601

**Comment by simonemainardi** [comment]
@Queeq I've just spent 1 hour trying to spot mem leaks but I can't find any. 

Could you please try with 
`valgrind --tool=memcheck --leak-check=yes ./ntopng`

and give us feedback? Thank you

**Comment by Queeq** [comment]
@simonemainardi 
Running under `valgrind` as you said didn't lead to freeze. However there is a significant difference in its behavior: it was showing like 1/500 of actual bandwidth and ~1/100 of number of flows and hosts. It was also consuming much less memory (~500 MB vs ~3 GB) and CPU (~100% instead of ~400%). It seemed like it was only able to run on one core and was missing a lot of data from nprobes. Maybe you could suggest other valgrind options?

**Comment by Queeq** [comment]
Attaching valgrind log, just in case
[ntopng_valgrind.log.txt](https://github.com/ntop/ntopng/files/293759/ntopng_valgrind.log.txt)

**Comment by simonemainardi** [comment]
@Queeq please attach the full ntopng command line used. I am not sure it's the same you reported above. Indeed, in the valgrind log I can find MySQL calls.

Also, does ntopng freezes while using the UI? Or it eats up all the ram regardless the use of the UI?

**Comment by Queeq** [comment]
Last time it was run as `valgrind --tool=memcheck --leak-check=yes --time-stamp=yes ntopng /etc/ntopng/ntopng.conf`

The ntopng.conf is the following:

```
-G=/var/tmp/ntopng.pid
--community
--http-port=:3000
--http-prefix="/ntopng"
--dns-mode=3
--max-num-flows=200000
--max-num-hosts=250000
--sticky-hosts=none
--dump-flows=mysql;localhost;ntopng;flows;ntopng;<cut>
--dump-hosts=all
-i=tcp://127.0.0.1:5556
-i=tcp://127.0.0.1:5557
-i=tcp://127.0.0.1:5558
-i=tcp://127.0.0.1:5559
-i=tcp://127.0.0.1:5562
-i=tcp://127.0.0.1:5564
-i=view:tcp://127.0.0.1:5557,tcp://127.0.0.1:5558,tcp://127.0.0.1:5559,tcp://127.0.0.1:5562
-i=view:tcp://127.0.0.1:5556,tcp://127.0.0.1:5557,tcp://127.0.0.1:5558,tcp://127.0.0.1:5559,tcp://127.0.0.1:5562,tcp://127.0.0.1:5564
```

Ntopng freeezes only after UI had been opened for some time. It could run for days without problems when UI is closed.

**Comment by simonemainardi** [comment]
@Queeq , since the issue only occurs when the UI is running, I think we can safely _exclude_ the following things:
- MySQL dump
- periodic lua activities (scripts/callbacks) and C methods invoked there

If you are not experiencing the errors: `WARNING: Script failure [/usr/share/ntopng/scripts/lua/get_grouped_hosts_data.lua][not enough memory]
` then we can also exclude that Lua is eating up too much memory.

Therefore, I argue there is some Lua-invoked C method that may be leaking and that is only invoked when actively browsing the UI.

For this reason I kindly encourage you to carefully browse the UI and report possible actions that may trigger (or even increase the speed) of the freezes. You were talking about the AS page. Are you experiencing this behavior even somewhere else?

**Comment by Queeq** [comment]
Well, actually it becomes frozen when these messages appear:

```
03/Jun/2016 19:09:01 [Lua.cpp:5063] WARNING: Script failure [/usr/share/ntopng/scripts/lua/get_grouped_hosts_data.lua][not enough memory]
03/Jun/2016 19:09:01 [Lua.cpp:5063] WARNING: Script failure [/usr/share/ntopng/scripts/lua/get_grouped_hosts_data.lua][not enough memory]
03/Jun/2016 19:09:01 [Lua.cpp:5063] WARNING: Script failure [/usr/share/ntopng/scripts/lua/get_grouped_hosts_data.lua][not enough memory]
03/Jun/2016 19:09:01 [Lua.cpp:5063] WARNING: Script failure [/usr/share/ntopng/scripts/lua/get_grouped_hosts_data.lua][not enough memory]
```

Active browsing is not needed, simply leaving the following pages open lead to freeze with the above messages:
- Hosts -> Autonomous Systems
- Hosts -> Countries

Keeping the following pages opened doesn't lead to freeze:
- Hosts -> Networks (I have no networks defined)
- Hosts -> MAC addresses (I have only one MAC  there - all zeros)
- Hosts -> Hosts (that's unexpected!)
- Flows

I'm sure all others are good too as I never experienced any issues with them. I mentioned here only those I've tested just now on v.2.3.160603

What I also noticed is that number of hosts (the one near the uptime on the lower right) grows indefinitely when UI is opened on the problematic pages. Even above the set limit. Looks like they are not getting cleared somehow. If I switch to another page - they are cleared. If this happened before it became frozen, I can use AS page for some more time.

**Comment by simonemainardi** [comment]
@Queeq any chance to have remote access (SSH + web) to the system? If affirmative, please send the details to mainardi ntop org

thank you

**Comment by Queeq** [comment]
@simonemainardi Sorry, the machine is located within a restricted segment of a corporate network. Let me know what kind of diagnostics you would like to see.

**Comment by simonemainardi** [comment]
Ok, let's proceed by trial and error. I've made additional changes. Please try again tomorrow with the latest nighty build and report. Thank you.

**Comment by Queeq** [comment]
Seems like v.2.3.160608 is slightly better. As if you made purging inactive hosts faster. However number of hosts is still growing. There is an impression it has not enough resources to purge them when UI is opened thus leading to memory overload and eventual freeze.

I tried setting host idle timeouts to 30 seconds together with Hosts Stats Update Frequency to 10 seconds, which made it better and number of hosts stopped growing indefinitely. I'll watch it for a bit more time and report back.

**Comment by simonemainardi** [comment]
@Queeq I was wondering how's it going after 1 day...

**Comment by Queeq** [comment]
I was not able to leave UI opened for long and haven't been watching it as I was traveling. Now as I've looked at it - it's frozen and ate all RAM and swap. There is this message in the log again:

```
08/Jun/2016 22:25:49 [Lua.cpp:4721] WARNING: Script failure [/usr/share/ntopng/scripts/callbacks/minute.lua][not enough memory]
```

It appeared after it had been running for about 10 hours. UI was closed. It had continued working though as there are some more messages regarding lost connection to MySQL later. Seems like only parts of ntopng don't respond (e.g. web-server).

**Comment by Queeq** [comment]
Oh, it wasn't ntopng who ate all memory (it consumed only around 4G), but redis-server. Became clear after I restarted ntopng (had to send SIGABRT as there was no reaction on SIGTERM).

**Comment by Queeq** [comment]
By the way, this is a real problem with redis as it consumes over 5 GB even after restart. Seems like it's packed with obsolete data from ntopng:

```
➜ redis-cli --bigkeys

# Scanning the entire keyspace to find biggest keys as well as
# average sizes per key type.  You can use -i 0.1 to sleep 0.1 sec
# per 100 SCAN commands (not usually needed).

[00.00%] Biggest hash   found so far 'ntopng.160519.hostkeys' with 928782 fields
[00.00%] Biggest string found so far 'ntopng.prefs.tcp://127.0.0.1:5558.speed' with 5 bytes
[00.00%] Biggest list   found so far 'ntopng.trace' with 33 items
[00.00%] Biggest hash   found so far 'ntopng.160604.hostkeys' with 5703634 fields
[00.00%] Biggest hash   found so far 'ntopng.160525.hostkeys' with 6613308 fields
[00.00%] Biggest string found so far 'ntopng.user.queeq.full_name' with 14 bytes
[13.70%] Biggest string found so far 'ntopng.user.admin.full_name' with 20 bytes
[27.40%] Biggest string found so far 'ntopng.user.queeq.password' with 32 bytes
[54.79%] Biggest string found so far 'ntopng.prefs.queeq.ifname' with 130 bytes

-------- summary -------

Sampled 73 keys in the keyspace!
Total key length in bytes is 2642 (avg len 36.19)

Biggest string found 'ntopng.prefs.queeq.ifname' has 130 bytes
Biggest   list found 'ntopng.trace' has 33 items
Biggest   hash found 'ntopng.160525.hostkeys' has 6613308 fields

53 strings with 611 bytes (72.60% of keys, avg size 11.53)
3 lists with 37 items (04.11% of keys, avg size 12.33)
0 sets with 0 members (00.00% of keys, avg size 0.00)
17 hashs with 43853399 fields (23.29% of keys, avg size 2579611.71)
0 zsets with 0 members (00.00% of keys, avg size 0.00)
```

**Comment by simonemainardi** [comment]
@Queeq thanks for the feedback. In order to verify this issue we should wait until you're able to leave the UI open on the AS page.

As per the redis memory consumption, you may want to try and disable dns caching and host dump. Also this may help: https://github.com/ntop/ntopng/blob/dev/doc/README.redis

**Comment by Queeq** [comment]
@simonemainardi Thanks for the redis hint.

As of now, UI has being opened for around 3 hours and is still running (v.2.3.160610). So the freeze issue is gone. However there are some other:
1. ntopng has eaten 5.5 GB of memory so far, and
2. it shows 630k active hosts (even though I had set the limit to 500k) and slowly rising.
3. AS throughput is incorrect (2-3 times less than actual).

As for points 1-2, I still have an impression that it is unable to clear old hosts.
As for point 3, I know what minimum throughput should be (some links are used solely for traffic of our AS). Ntopng used to show correct values until recent changes (not sure when it began, maybe a week or two ago). This issue seems to be more true for combined interfaces (views). It happens that a view that combines 6 interfaces shows less traffic than a view that combines 3. And for some ASes a single interface may show more traffic than a view where it's included. There were no changes in sampling configs.

Could it be that some of the commits that were aimed to resolve the freezing issue lead to such regression?

**Comment by simonemainardi** [comment]
@Queeq thanks for reporting and sorry for the late feedback.

As we are in the process of re-designing interface views, we thought we could just pick the first physical interface in the grouped statistics. Now with the new commit we have added all the physical interfaces to the count.

You should be able to update and get the proper values. Also, as out-of-memory errors are gone I also think it is safe to close this issue. Shall you have additional feedback, please let us know.

**Comment by simonemainardi** [comment]
The latest commit fixed the issue. Closing.
